prov/tcp/src/tcpx_ep.c

```diff
@@ -405,6 +405,84 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
 	return FI_SUCCESS;
 }
 
+/* when src_addr is NULL, addr_format and addrlen are ignored */
+static int tcpx_pep_create_listen_socket(struct tcpx_pep *pep, void *src_addr,
+					 uint32_t addr_format, size_t addrlen)
+{
+	struct addrinfo hints, *result, *iter;
+	char sa_ip[INET_ADDRSTRLEN] = {0};
+	char sa_port[NI_MAXSERV] = {0};
+	int ret;
+
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_flags = AI_PASSIVE;
+
+	if (src_addr) {
+		switch (addr_format) {
+		case FI_SOCKADDR:
+		case FI_SOCKADDR_IN:
+		case FI_SOCKADDR_IN6:
+			ret = getnameinfo(src_addr,
+					  (socklen_t) addrlen,
+					  sa_ip, INET_ADDRSTRLEN,
+					  sa_port, NI_MAXSERV, 0);
+			if (ret) {
+				FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+					"pep initialization failed\n");
+				return ret;
+			}
+			break;
+		default:
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"invalid source address format\n");
+			return -FI_EINVAL;
+
+		}
+		ret = getaddrinfo(sa_ip, sa_port, &hints, &result);
+	} else {
+		ret = getaddrinfo("localhost", NULL, &hints, &result);
+	}
+
+	if (ret) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"getaddrinfo failed");
+		return -FI_EINVAL;
+	}
+
+	for (iter = result; iter; iter = iter->ai_next) {
+		pep->sock = ofi_socket(iter->ai_family, iter->ai_socktype,
+				       iter->ai_protocol);
+		if (pep->sock == INVALID_SOCKET)
+			continue;
+
+		ret = tcpx_setup_socket(pep->sock);
+		if (ret) {
+			ofi_close_socket(pep->sock);
+			pep->sock = INVALID_SOCKET;
+			continue;
+		}
+
+		if (bind(pep->sock, result->ai_addr,
+			 (socklen_t) result->ai_addrlen)) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: %s\n", strerror(errno));
+			ofi_close_socket(pep->sock);
+			pep->sock = INVALID_SOCKET;
+		} else {
+			break;
+		}
+	}
```

Comment:
I know this is just moving code, but this looks more complex than needed.  Why doesn't the code simply do something like this:
```
if (src) {
    pep->sock = ofi_socket(src->sa_family, SOCK_STREAM, 0);
    ofi_bind(pep->sock, src, ofi_addrlen(src));
} else {
    pep->sock = ofi_socket(AF_INET, SOCK_STREAM, 0);
}
```
In theory, we've already gone through the process of iterating over the address list as part of fi_getinfo().  So, if a src address has been set (and it's possible that it's always been set), then we just use it.  If the src has not been set, then we should look at why fi_getinfo() did not set it, and see if it should have.  If src should always be set, then we should just fail the call.